src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,307 @@
+import { produce } from 'immer'
+import { EM_TOKEN, DUMMY_TOKEN } from '../constants.js'
+
+const EDIT_TIME_MAX = 7200 // time diff limit in second for replacing descendants by ancestor
+
+/**
+   * finds tree deepeset common subcontext for given a given path
+   * @param {Object} tree nested object representing tree
+   * @param {string[]} context array of string representing path
+   * @returns {string[]} common subcontext
+   */
+const findTreeDeepestSubcontext = (tree, context, index = 0) => {
+  if (context.length === 0 && index === 0) return { node: tree, path: [] }
+  const node = tree[context[index]]
+  return node
+    ? findTreeDeepestSubcontext(node, context, index + 1)
+    : { node: tree, path: context.slice(0, index) }
+}
+
+/**
+   * finds all the desecendant for a given context of a specific node
+   * @param {Object} parentNode nested object representing tree
+   * @param {string[]} startingPath context of the node whose descendants needs to be returned
+   * @returns {Object[]} array of descendant object
+   */
+export const findTreeDescendants = (tree, startingPath) => {
+  const node = startingPath && startingPath.length > 0
+    ? _.get(tree, startingPath)
+    : tree
+  return !node ? []
+    : node.leaf ? [{ ...node }]
+      : _.flatMap(Object.keys(node).map(
+        child => findTreeDescendants(node[child])
+      ))
+}
+
+/**
+   * finds closest ancestor that has multiple children nodes
+   * @param {object} tree nested object representing tree
+   * @param {number} [minChildren=2] mininum no of children
+   * @param {string[]} context array of string representing path
+   * @returns {Object} closest ancestor node with multiple children
+   */
+const findClosestSharedAncestor = (tree, context, minChildren = 2, index = 0, closestAncestor = { node: null, path: [] }) => {
+  const node = tree[context[index]]
+  return node
+    ? findClosestSharedAncestor(
+      node,
+      context,
+      minChildren,
+      index + 1,
+      (!node.leaf && Object.keys(node).length >= minChildren)
+        ? { node, path: context.slice(0, index + 1) }
+        : closestAncestor
+    )
+    : closestAncestor
+}
+
+/*
+  Example for nodeAdd
+
+  A.B.C.D.E.F
+  A.B.C.G.H.I
+
+  If we want to add A.B.C.K.L , then ['A','B','C',''] will be oldPath and ['A','B','C','K','L'] will be newPath.
+
+  So node add takes ['A','B','C'] as oldPath and ['K','L'] as addedPath and creates ['A','B','C',''] and ['A','B','C','K','L']
+  as second and third parameter for nodeChange
+*/
+
+/**
+   * Manually adding node to the existing tree object by mutating. (Uses special case of nodeChange)
+   * @param {object} tree nested object representing tree
+   * @param {string[]} oldPath array of string representing path
+   * @param {string[]} addedPath array of string representing path
+   *
+*/
+const nodeAdd = (tree, oldPath, addedPath) => nodeChange(tree, contextOf(oldPath).concat({ value: DUMMY_TOKEN }), oldPath.concat(addedPath))
```

Comment:
Here's another lesson about encapsulation I encourage you to think about.

When designing the signature for a function, ask yourself the questions:

- What is the minimum the caller needs to know to call this function?
- What is a natural expression of the inputs of this function?

Intuitively, the inputs for adding a node should simply be the tree and the node to add. The signature of `(tree, oldPath, addedPath) => ...` expects the caller to know something about `oldPath` and `addedPath` which is odd because it is related to the internal implementation of `nodeAdd` and not to the semantics of the caller.

Let me know if this makes sense!

Reply:
Hi Raine. It makes perfect sense. The problem here is `nodeAdd` is just a special case of `nodeChange` . So it needs `oldPath` that is the already available concat with `{value:""}` in the tree and `newPath` that is concat of `oldPath` and new branch added to that node.

Example 

  A.B.C.D.E.F
  A.B.C.G.H.I

  If we want to add `A.B.C.K.L` , then `['A','B','C','']` will be `oldPath` and `['A','B','C','K','L']` will be 
  `newPath`.

  So node add takes ['A','B','C'] as `oldPath` and ['K','L'] as `addedPath` and creates `['A','B','C','']` 
  and `['A','B','C','K','L']` to pass them as second and third parameter for `nodeChange`

Reply:
In 2/3 cases (Line 260 and 272), `oldPath` and `addedPath` appear to be derivable from `newPath`. The third case (Line 267) is very close. I am looking for a way to have all three of them use the same interface.

What about this?

```js
const nodeAdd = (tree, newPath) => 
  nodeChange(
    tree, 
    newPath.concat({ value: DUMMY_TOKEN }), 
    newPath
  )
```

Reply:
I understand your perspective but `addedPath` is not always derivable from `oldPath`. It's just in the case of line `260` and `272` that `oldPath` and `addedPath` are of same length. But in case of line `267`  the `addedPath` is longer path.

Reply:
Okay, just wanted to make sure. I think the arguments can still be derived from `commonPath`, but that would add a redundant call to `findTreeDeepestSubcontext` within `nodeMove`. I can live with this as-is for now. Thanks for engaging.